File: Java/496. Next Greater Element I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack<>();
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for (int n: nums2) {
            while (!stack.isEmpty() && n > stack.peek()) {
                map.put(stack.pop(), n);
            }
            stack.push(n);
        }
        
        for (int i=0; i<nums1.length; i++) {
            nums1[i] = map.getOrDefault(nums1[i], -1);
        }
        
        return nums1;
    }
}

File: Python3/496. Next Greater Element I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ref = {}
        stack = deque()
        
        for n in nums2:
            # case 1: found the next greater element
            while stack and n > stack[-1]:
                ref[stack.pop()] = n
                
            # case 2: the stack is empty
            # case 3: descending slope
            stack.append(n)
            
        for i, n in enumerate(nums1):
            nums1[i] = ref.get(n, -1)
                
        return nums1            
            
                